orttraining/orttraining/training_ops/cpu/tensor/gather_grad.cc

```diff
@@ -20,14 +21,14 @@ ONNX_OPERATOR_KERNEL_EX(
                                     DataTypeImpl::GetTensorType<int64_t>()}),
     GatherGrad);
 
-#define TYPED_GRAD_FUNCTION_CALL(T)                                      \
-  if (T_type == DataTypeImpl::GetType<T>()) {                            \
-    if (Tind_type == DataTypeImpl::GetType<int32_t>()) {                 \
-      return ComputeImpl<T, int32_t>(data_shape, indices, grad, output); \
-    }                                                                    \
-    if (Tind_type == DataTypeImpl::GetType<int64_t>()) {                 \
-      return ComputeImpl<T, int64_t>(data_shape, indices, grad, output); \
-    }                                                                    \
+#define TYPED_GRAD_FUNCTION_CALL(T, tp)                                      \
```

Comment:
Would a template function provide a safer type checking for T and tp?

Reply:
Not necessarily safer but faster. That being said, the point of this PR isn't to optimize/clean up code but to replace OpenMP with the new ThreadPool abstractions. Thanks!